gh-90016: Deprecate default sqlite3 adapters and converters (#94276)
python/cpython: Doc/library/sqlite3.rst

.. literalinclude:: ../includes/sqlite3/converter_point.py

> Default adapters and converters (deprecated)
> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Reviewer: ```suggestion
.. _sqlite3-default-converters:

Default adapters and converters (deprecated)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Add ref target to link directly in NEWS and What's New.

> .. note::

Reviewer: ```suggestion
.. warning::
```

Wouldn't it make sense to upgrade this to an actual warning, since it refers to deprecated behavior and triggers a warning at run time in Python itself, rather than just a note? (As sidenote, this would also help it look less repetitive vis a vi the note in the paragraph after.)

Reviewer 2: @CAM-Gerlach: Are warnings used for deprecations elsewhere, or are they generally reserved for security warnings and the like? (Honest question, I don't know the answer.)

Author: I'm not sure. My first reaction is that the warning directive is a little bit over the top here. I'm leaning against keeping `note`.

Reviewer: The standard way for the stdlib reference is using the ``deprecated-removed` directive, which is also used here below. I'm not sure if there's any real standard for "how-to guide" material like this, though personally I'd lean toward drawing more attention to features being deprecated rather than less, and risking people glancing over it. However, it ultimately appears to be at the discretion of the doc authors and module maintainers, i.e. @erlend-aasland .

Author: As there is no precedence for using the `warning` directive for these things, I'll keep it as a `note`. Note that `note` is used in a handful of places in `Doc/library`, where deprecation is mentioned in some kind of way.

   The default adapters and converters are deprecated as of Python 3.12.

>    Use the :ref:`sqlite3-adapter-converter-recipes`

Author: ```suggestion
   Use instead the :ref:`sqlite3-adapter-converter-recipes`
```

or:

```suggestion
   Instead, use the :ref:`sqlite3-adapter-converter-recipes`
```

Reviewer: Second one, as you presumably found out; the first one is not the right word order in English (though it is in a number of other languages). Just for reference, another valid alternative is

```rst
Use the :ref:`sqlite3-adapter-converter-recipes` instead.
```

Thought I think I slightly prefer the wording you went with, since it sounds a bit gentler.

>    Instead, use the :ref:`sqlite3-adapter-converter-recipes`,

Reviewer: ```suggestion
   Instead, use the :ref:`sqlite3-adapter-converter-recipes`
```

I don't think we need a comma here

   and tailor them to your needs.

> The deprecated default adapters and converters consists of:

Reviewer: ```suggestion
The deprecated default adapters and converters consist of:
```

Grammar: plural agreement

> * Adapt :class:`datetime.date` objects to :class:`strings <str>` in ISO format.
> * Adapt :class:`datetime.datetime` objects to strings ISO format.
> * Convert :ref:`declared <sqlite3-converters>` "date" types to ``datetime.date``
>   objects.
> * Convert declared "timestamp" types to ``datetime.datetime`` objects.

Reviewer: ```suggestion
* An adapter for :class:`datetime.date` objects to :class:`strings <str>` in
  `ISO 8601 <https://en.wikipedia.org/wiki/ISO_8601>`_ format.
* An adapter for :class:`datetime.datetime` objects to strings in ISO 8601 format.
* A converter for :ref:`declared <sqlite3-converters>` "date" types to
  :class:`datetime.date` objects.
* A converter for declared "timestamp" types to :class:`datetime.datetime` objects.
```

* Explicitly specify and link ISO 8601 for clarity
* Fix grammar/agreement (tip: try inserting the bullet items after the colon—do they make grammatical sense as a sentence? E.g. "The adapters and converters consist of: adapt date objects to strings in ISO format." vs. "The adapters and converters consist of: an adapter for date objects to strings in ISO format.")
* Fix typo (missing word) in second item
* Use consistent `:class:` role/formatting for classes

>   Fractional parts will be truncated to 6 digits (microsecond precision).

Reviewer: Just to be clear—its truncation, not rounding?

Author: Yep, truncation.

.. literalinclude:: ../includes/sqlite3/converter_point.py

Default adapters and converters (deprecated)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. note::

   The default adapters and converters are deprecated as of Python 3.12.
   Use the :ref:`sqlite3-adapter-converter-recipes`
   and tailor them to you need.

The deprecated default adapters and converters consists of:

> * Adapt :class:`datetime.date` objects to ISO format :class:`strings <str>`.

Reviewer 3: I think it would be a little better to say "strings in ISO format". Ditto for the line below.

Author: Thanks; addressed in 2ba83f60ad511236fe666916bf9d541f3cd988c7

* Adapt :class:`datetime.datetime` objects to ISO format :class:`strings <str>`.
* Convert :ref:`declared <sqlite3-converters>` "date" types to ``datetime.date``
  objects.
* Convert declared "timestamp" types to ``datetime.datetime`` objects.
  If a timestamp stored in SQLite has a fractional part longer than 6

>   numbers, its value will be truncated to microsecond precision.

Reviewer 3: Should be "digits" rather than "numbers", but I think this sentence would be clearer if both halves used the same unit of measurement, because I can't remember how many digits "microsecond precision" is. Maybe just say "Timestamps stored in SQLite will be truncated to 6 digits (microsecond precision)"?

Author: Totally agree. I just copied the previous sentence. Your suggestion is definitely an improvement.

Author: PTAL. Addressed in 2ba83f60ad511236fe666916bf9d541f3cd988c7.

Reviewer 3: Looks good to me

   offsets in timestamps, either leave converters disabled, or register an
   offset-aware converter with :func:`register_converter`.

.. versionchanged:: 3.12

   Deprecated default adapters and converters.

> .. deprecated:: 3.12

Reviewer: ```suggestion
.. deprecated:: 3.12
```

This is kinda repetitive and redundant, since it is immediately followed by a more specific and appropriate directive repeating the same thing, and the section heading, text and admonition also state this. If you want to be absolutely explicit about what is deprecated, you could do

```suggestion
.. deprecated:: 3.12

   Default adapters and converters.
```
